File: ci-tests/tests/testers/hoard_status.py
import os
from pathlib import Path
import uuid

from .hoard_tester import HoardTester


DEFAULT_CONTENT = "default text"
CHANGED_CONTENT = "changed text"
OTHER_CONTENT = "other text"


class StatusCommandTester(HoardTester):
    def setup(self):
        self.env["HOARD_LOG"] = "info"
        self.run_hoard("validate")
        self.local_uuid = str(uuid.uuid4())

        self.run_hoard("validate")
        self.remote_uuid = str(uuid.uuid4())

        self.args = ["--force"]

    def _setup_no_changes(self):
        path = Path.home().joinpath("unchanged.txt")
        self.uuid = self.local_uuid
        self._write_file(path, DEFAULT_CONTENT, is_binary=False)
        self.targets = ["no_changes"]
        self.run_hoard("backup")

    def _setup_local_changes(self):
        path = Path.home().joinpath("local.txt")
        self.uuid = self.remote_uuid
        self._write_file(path, DEFAULT_CONTENT, is_binary=False)
        self.targets = ["local_changes"]
        self.run_hoard("backup")
        self.uuid = self.local_uuid
        self.run_hoard("restore")
        self._write_file(path, CHANGED_CONTENT, is_binary=False)

    def _setup_remote_changes(self):
        path = Path.home().joinpath("remote.txt")
        self.uuid = self.local_uuid
        self._write_file(path, DEFAULT_CONTENT, is_binary=False)
        self.targets = ["remote_changes"]
        self.run_hoard("backup")

        self.uuid = self.remote_uuid
        self.run_hoard("restore")
        self._write_file(path, CHANGED_CONTENT, is_binary=False)
        self.run_hoard("backup")

        self._write_file(path, DEFAULT_CONTENT, is_binary=False)

    def _setup_mixed_changes(self):
        path = Path.home().joinpath("mixed.txt")
        self.uuid = self.local_uuid
        self._write_file(path, DEFAULT_CONTENT, is_binary=False)
        self.targets = ["mixed_changes"]
        self.run_hoard("backup")

        self.uuid = self.remote_uuid
        self._write_file(path, CHANGED_CONTENT, is_binary=False)
        self.run_hoard("backup")

        self._write_file(path, OTHER_CONTENT, is_binary=False)

    def _setup_unexpected_changes(self):
        path = Path.home().joinpath("unexpected.txt")
        hoard_path = self.data_dir_path().joinpath("hoards", "unexpected_changes")
        self._write_file(path, DEFAULT_CONTENT, is_binary=False)
        self.uuid = self.local_uuid
        self.targets = ["unexpected_changes"]
        self.run_hoard("backup", capture_output=True)
        self._write_file(hoard_path, CHANGED_CONTENT, is_binary=False)

    def run_test(self):
        self.reset(config_file="hoard-status-config.toml")
        self._setup_no_changes()
        self._setup_local_changes()
        self._setup_remote_changes()
        self._setup_mixed_changes()
        self._setup_unexpected_changes()

        self.uuid = self.local_uuid
        self.args = []
        self.targets = []
        result = self.run_hoard("status", capture_output=True)

        assert b"no_changes: up to date\n" in result.stdout
        assert b"local_changes: modified locally -- sync with `hoard backup local_changes`\n" in result.stdout, f"got \"{result.stdout}\""
        assert b"remote_changes: modified remotely -- sync with `hoard restore remote_changes`\n" in result.stdout, f"got \"{result.stdout}\""
        assert b"mixed_changes: mixed changes -- manual intervention recommended (see `hoard diff mixed_changes`)\n" in result.stdout, f"got \"{result.stdout}\""
        assert b"unexpected_changes: unexpected changes -- manual intervention recommended (see `hoard diff unexpected_changes`)\n" in result.stdout, f"got \"{result.stdout}\""

File: tests/hoard_status.rs
mod common;

use common::tester::Tester;
use hoard::command::Command;
use tokio::fs;

const DEFAULT_CONTENT: &str = "default text";
const CHANGED_CONTENT: &str = "changed text";
const OTHER_CONTENT: &str = "other text";

const HOARD_NO_CHANGES: &str = "no_changes";
const HOARD_LOCAL_CHANGES: &str = "local_changes";
const HOARD_REMOTE_CHANGES: &str = "remote_changes";
const HOARD_MIXED_CHANGES: &str = "mixed_changes";
const HOARD_UNEXPECTED_CHANGES: &str = "unexpected_changes";

const STATUS_TOML: &str = r#"
exclusivity = [
    ["unix", "windows"]
]

[envs]
[envs.windows]
    os = ["windows"]
[[envs.windows.env]]
    var = "HOARD_TMP"
[envs.unix]
    os = ["linux", "macos"]
[[envs.unix.env]]
    var = "HOME"

[hoards]
[hoards.no_changes]
    "unix"    = "${HOME}/unchanged.txt"
    "windows" = "${HOARD_TMP}/unchanged.txt"
[hoards.local_changes]
    "unix"    = "${HOME}/local.txt"
    "windows" = "${HOARD_TMP}/local.txt"
[hoards.remote_changes]
    "unix"    = "${HOME}/remote.txt"
    "windows" = "${HOARD_TMP}/remote.txt"
[hoards.mixed_changes]
    "unix"    = "${HOME}/mixed.txt"
    "windows" = "${HOARD_TMP}/mixed.txt"
[hoards.unexpected_changes]
    "unix"    = "${HOME}/unexpected.txt"
    "windows" = "${HOARD_TMP}/unexpected.txt"
"#;

async fn setup_no_changes(tester: &Tester) {
    let path = tester.home_dir().join("unchanged.txt");
    fs::write(&path, DEFAULT_CONTENT)
        .await
        .expect("writing to file should succeed");
    tester.use_local_uuid().await;
    tester
        .expect_forced_command(Command::Backup {
            hoards: vec![HOARD_NO_CHANGES.parse().unwrap()],
        })
        .await;
}

async fn setup_local_changes(tester: &Tester) {
    let path = tester.home_dir().join("local.txt");
    fs::write(&path, DEFAULT_CONTENT)
        .await
        .expect("writing to file should succeed");
    tester.use_remote_uuid().await;
    tester
        .expect_forced_command(Command::Backup {
            hoards: vec![HOARD_LOCAL_CHANGES.parse().unwrap()],
        })
        .await;
    tester.use_local_uuid().await;
    tester
        .expect_forced_command(Command::Restore {
            hoards: vec![HOARD_LOCAL_CHANGES.parse().unwrap()],
        })
        .await;
    fs::write(&path, CHANGED_CONTENT)
        .await
        .expect("writing to file should succeed");
}

async fn setup_remote_changes(tester: &Tester) {
    let path = tester.home_dir().join("remote.txt");
    fs::write(&path, DEFAULT_CONTENT)
        .await
        .expect("writing to file should succeed");
    tester.use_local_uuid().await;
    tester
        .expect_forced_command(Command::Backup {
            hoards: vec![HOARD_REMOTE_CHANGES.parse().unwrap()],
        })
        .await;
    tester.use_remote_uuid().await;
    tester
        .expect_forced_command(Command::Restore {
            hoards: vec![HOARD_REMOTE_CHANGES.parse().unwrap()],
        })
        .await;
    fs::write(&path, CHANGED_CONTENT)
        .await
        .expect("writing to file should succeed");
    tester
        .expect_forced_command(Command::Backup {
            hoards: vec![HOARD_REMOTE_CHANGES.parse().unwrap()],
        })
        .await;
    fs::write(&path, DEFAULT_CONTENT)
        .await
        .expect("writing to file should succeed");
}

async fn setup_mixed_changes(tester: &Tester) {
    let path = tester.home_dir().join("mixed.txt");
    tester.use_local_uuid().await;
    fs::write(&path, DEFAULT_CONTENT)
        .await
        .expect("writing to file should succeed");
    tester
        .expect_forced_command(Command::Backup {
            hoards: vec![HOARD_MIXED_CHANGES.parse().unwrap()],
        })
        .await;
    tester.use_remote_uuid().await;
    fs::write(&path, CHANGED_CONTENT)
        .await
        .expect("writing to file should succeed");
    tester
        .expect_forced_command(Command::Backup {
            hoards: vec![HOARD_MIXED_CHANGES.parse().unwrap()],
        })
        .await;
    tester.use_local_uuid().await;
    fs::write(&path, OTHER_CONTENT)
        .await
        .expect("writing to file should succeed");
}

async fn setup_unexpected_changes(tester: &Tester) {
    let path = tester.home_dir().join("unexpected.txt");
    let hoard_path = tester.data_dir().join("hoards").join("unexpected_changes");
    tester.use_local_uuid().await;
    fs::write(&path, DEFAULT_CONTENT)
        .await
        .expect("writing to file should succeed");
    tester
        .expect_forced_command(Command::Backup {
            hoards: vec![HOARD_UNEXPECTED_CHANGES.parse().unwrap()],
        })
        .await;
    fs::write(&hoard_path, CHANGED_CONTENT)
        .await
        .expect("writing to file should succeed");
}

#[tokio::test]
async fn test_hoard_status() {
    let tester = Tester::new(STATUS_TOML).await;
    setup_no_changes(&tester).await;
    setup_local_changes(&tester).await;
    setup_remote_changes(&tester).await;
    setup_mixed_changes(&tester).await;
    setup_unexpected_changes(&tester).await;

    tester.use_local_uuid().await;
    tester.expect_command(Command::Status).await;

    tester.assert_has_output("no_changes: up to date\n");
    tester.assert_has_output(
        "local_changes: modified locally -- sync with `hoard backup local_changes`\n",
    );
    tester.assert_has_output(
        "remote_changes: modified remotely -- sync with `hoard restore remote_changes`\n",
    );
    tester.assert_has_output("mixed_changes: mixed changes -- manual intervention recommended (see `hoard diff mixed_changes`)\n");
    tester.assert_has_output("unexpected_changes: unexpected changes -- manual intervention recommended (see `hoard diff unexpected_changes`)\n");
}
